lib/logstash/outputs/webhdfs.rb

```diff
@@ -0,0 +1,282 @@
+# encoding: utf-8
+require "logstash/namespace"
+require "logstash/outputs/base"
+require "stud/buffer"
+
+# Summary: Plugin to send logstash events to to files in HDFS via webhdfs 
+# restapi.  
+#
+# This plugin only has a mandatory dependency on the webhdfs gem from 
+# Kazuki Ohta and TAGOMORI Satoshi (@see: https://github.com/kzk/webhdfs).
+# Optional dependencies are zlib and snappy gem. 
+# No jars from hadoop are needed, thus reducing configuration and compatibility
+# problems.
+#
+# If you get an error like:
+#
+#     Max write retries reached. Exception: initialize: name or service not known {:level=>:error}
+#
+# make sure, that the hostname of your namenode is resolvable on the host running logstash. When creating/appending
+# to a file, webhdfs somtime sends a 307 TEMPORARY_REDIRECT with the HOSTNAME of the machine its running on.
+#
+# USAGE:
+# This is an example of logstash config:
+#
+# [source,ruby]
+# ----------------------------------
+# webhdfs {
+#   server => "127.0.0.1:50070"         # (required)
+#   path => "/user/logstash/dt=%{+YYYY-MM-dd}/logstash-%{+HH}.log"  # (required)
+#   user => "hue"                       # (optional)
+#   message_format => "%{@source_host}" # (optional)
+#   idle_flush_time => 10               # (optional)
+#   flush_size => 50                    # (optional)
+#   open_timeout => 15                  # (optional)
+#   read_timeout => 15                  # (optional)
+#   use_httpfs => true                  # (optional)
+#   retry_interval => 1                 # (optional)
+#   retry_times => 3                    # (optional)
+#   compress => "snappy"                # (optional)
+#   remove_at_timestamp => false        # (optional)
+# }
+# ----------------------------------
+#
+# Author: Bjoern Puttmann <b.puttmann@dbap.de> - dbap GmbH, Münster, Germany.
+
+class LogStash::Outputs::WebHdfs < LogStash::Outputs::Base
+  include Stud::Buffer
+
+  config_name "webhdfs"
+  milestone 1
+
+  if RUBY_VERSION[0..2] == '1.8'
+    MAGIC = "\x82SNAPPY\x0"
+  else
+    MAGIC = "\x82SNAPPY\x0".force_encoding Encoding::ASCII_8BIT
+  end
+  DEFAULT_VERSION = 1
+  MINIMUM_COMPATIBLE_VERSION = 1
+
+  # The server name and port for webhdfs/httpfs connections.
+  config :server, :validate => :string, :required => true
```

Comment:
Let's make these settings separate, "host" and "port"

```ruby
config :host, :validate => :string, :required => true
config :port, :validate => :number, :default => ....
```

(The default for port assumes there's a default port, otherwise make it required)